peps/pep-0757.rst

```diff
@@ -0,0 +1,402 @@
+PEP: 757
+Title: C API to import-export Python integers
+Author: Sergey B Kirpichev <skirpichev@gmail.com>,
+        Victor Stinner <vstinner@python.org>
+PEP-Delegate: C API Working Group
+Status: Draft
+Type: Standards Track
+Created: 13-Sep-2024
+Python-Version: 3.14
+
+.. highlight:: c
+
+
+Abstract
+========
+
+Add a new C API to import and export Python integers, ``int`` objects:
+especially ``PyLongWriter_Create()`` and ``PyLong_AsDigitArray()``
+functions.
+
+
+Rationale
+=========
+
+Projects such as `gmpy2 <https://github.com/aleaxit/gmpy>`_, `SAGE
+<https://www.sagemath.org/>`_ and `Python-FLINT
+<https://github.com/flintlib/python-flint>`_ access directly Python
+"internals" (the ``PyLongObject`` structure) or use an inefficient
+temporary format (hex strings for Python-FLINT) to import and
+export Python ``int`` objects.  The Python ``int`` implementation
+changed in Python 3.12 to add a tag and "compact values".
+
+In the 3.13 alpha 1 release, the private undocumented ``_PyLong_New()``
+function had been removed, but it is being used by these projects to
+import Python integers. The private function has been restored in 3.13
+alpha 2.
+
+A public efficient abstraction is needed to interface Python with these
+projects without exposing implementation details. It would allow Python
+to change its internals without breaking these projects.
+
+
+Specification
+=============
+
+Layout API
+----------
+
+API::
+
+    typedef struct PyLongLayout {
+        // Bits per digit
+        uint8_t bits_per_digit;
+
+        // Digit size in bytes
+        uint8_t digit_size;
+
+        // Digits order:
+        // * 1 for most significant digit first (big endian)
+        // * -1 for least significant digit first (little endian)
+        int8_t digits_order;
+
+        // Endian:
+        // * 1 for most significant byte first (big endian)
+        // * -1 for least significant byte first (little endian)
+        int8_t endian;
+    } PyLongLayout;
+
+    PyAPI_FUNC(const PyLongLayout*) PyLong_GetNativeLayout(void);
```

Comment:
I still worry that this structure has only two useful field: bits_per_digit and digit_size.  And this already present in PyLong_GetInfo().

Reply:
Ok, I added it as an "open question" to the PEP.